module9/ImageSearch/ImageSearch/Models/ImageSearchStore.swift

```diff
@@ -0,0 +1,229 @@
+    resetLocalState()
+
+    guard let newSearchURL = getSearchURL(query: query ) else {
+      searchLoadingState = .error
+      return
+    }
+    searchLoadingState = .loadingSearch
+    currentQuery = query
+
+    do {
+      let apiRequest = try getPexelsAPIRequest(for: newSearchURL)
+      self.urlSessionLoader.fetchDataFromURL(urlRequest: apiRequest) { data, response, error in
+        self.processSearchFetch(data: data, response: response, error: error)
+      }
+    } catch {
+      searchLoadingState = .error
+    }
+  }
+
+  public func performNextPageSearch() {
+    guard let nextPageURL = nextPageURL else {
+      searchLoadingState = .error
+      return
+    }
+    searchLoadingState = .loadingSearch
+
+    do {
+      let apiRequest = try getPexelsAPIRequest(for: nextPageURL)
+      self.urlSessionLoader.fetchDataFromURL(urlRequest: apiRequest) { data, response, error in
+        self.processSearchFetch(data: data, response: response, error: error)
+      }
+    } catch {
+      searchLoadingState = .error
+    }
+  }
+
+  // Internal Helpers
+
+  func loadPlexelsAPIKey(fileName: String) -> String {
+    guard let filePath = Bundle.main.path(forResource: fileName, ofType: "plist") else {
+      // TODO own error enum for error handling
+      fatalError("loadPlexelsAPIKey: Couldn't find file '\(fileName).plist'.")
+    }
+    let plist = NSDictionary(contentsOfFile: filePath)
+    guard let value = plist?.object(forKey: "API_KEY") as? String else {
+      fatalError("Couldn't find key 'API_KEY' in '\(fileName).plist'.")
+    }
+    return value
+  }
+
+  func resetLocalState() {
+    currentQuery = nil
+    totalResults = nil
+    currentPage = nil
+    imageResults = []
+    nextPageURL = nil
+  }
+
+  func processSearchFetch(data: Data?, response: URLResponse?, error: Error?) {
+    if let data = data, let response = response as? HTTPURLResponse {
+      if response.statusCode != 200 {
+        self.searchLoadingState = .error // weak referecnce for self?
+      }
+
+      do {
+        let decoder = JSONDecoder()
+        let imageSearchResponse = try decoder.decode(ImageSearchResponse.self, from: data)
+        self.updateLocalStateWithSearchResponse(searchResponse: imageSearchResponse) // weak reference for self?
+
+      } catch let error {
+        print("Error decoding response \(String(describing: error))")
+        self.searchLoadingState = .error // weak referecnce for self?
+      }
+    } else {
+      print(
+        "Contents fetch failed: " +
+        "\(error?.localizedDescription ?? "Unknown error")")
+      self.searchLoadingState = .error // weak referecnce for self?
+    }
+  }
+
+  func updateLocalStateWithSearchResponse(searchResponse: ImageSearchResponse) {
+    imageResults.append(contentsOf: searchResponse.images)
+    totalResults = searchResponse.totalResults
+    currentPage = searchResponse.page
+    nextPageURL = URL(string: searchResponse.nextPage)
+    searchLoadingState = .loadedSearch
```

Comment:
OK this was really helpful. So from a best practices point of view, with this view-model style architecture, when I'm updating the view-model published state vars of the view model based on subscriptions to other objects (via combine) its prolly safest to wrap these view-model state vars in DispatchQueue.main ? That way we dont have to worry about the specifics of which threads various helper objects are publishing their updates on back to the view model?

Reply:
As I was working on something else, I was reminded of another, perhaps cleaner, way to do this. You make a main actor function that all it does is do whatever you're supposed to do to update the UI.

````
@MainActor
func updateLoadingState() {
  searchLoadingState = .loadedSearch
}
```
The @MainActor decoration says that this function must ALWAYS run on the main thread. Then you call it with an `await` in your `async` code. I didn't pass parameters, but you can certainly do that as well so that this update function would be more general.

````
<your async code running>
await updateLoadingState()
````